### PR TITLE
feat: ACNA-3098 - allow action invocation of ES Modules / Typescript

### LINKED ISSUE #42: skip uninteresting files during debugging
https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_skipping-uninteresting-code

Use the fixture below to test, then:
1. add to the template here https://github.com/adobe/generator-aio-app/blob/master/generators/add-vscode-config/index.js
2. update this fixture code https://github.com/adobe/generator-aio-app/blob/master/test/__fixtures__/add-vscode-config/launch.json
3. Update the docs at https://github.com/AdobeDocs/app-builder/pull/333

### LINKED ISSUE #69: Allow Action invocation of ES Modules
**Is your feature request related to a problem? Please describe.**
I am writing AIO actions in TypeScript and encountering issues with ES modules being used where CommonJS modules are expected after the build process. This results in runtime errors, and there are no aio app dev options like --no-build to circumvent this issue.

**Describe the solution you'd like**
I would like the build process to ensure compatibility with CommonJS and ES Modules. Specifically, this involves:

Using dynamic import() for ES modules within an async function.

**Describe alternatives you've considered**
Use Babel to Transpile ES Modules to CommonJS: Adds complexity and additional build steps, potentially causing issues with module resolution.

**Additional context**
Example error message: Error loading action function: require() of ES Module ... not supported. Instead change the require of ... to a dynamic import() which is available in all CommonJS modules.

### PR DESCRIPTION
**Pull request by shazron**
closes #69
closes #42

Enables ESM and TypeScript support for action code, via webpack.

## Pre-requisite PRs

- [x] https://github.com/adobe/aio-lib-runtime/pull/192

## Related PRs

- [x] https://github.com/adobe/generator-aio-app/pull/262
- [x] https://github.com/AdobeDocs/app-builder/pull/350

## TODO

- [x] add an action file watcher so files can be rebuilt (similar to `aio app run`). See [this](https://github.com/adobe/aio-cli-plugin-app/blob/master/src/lib/actions-watcher.js). This may introduce latency
- [X] ~remove the use of the magic string for the built action code in `dist`, we may have to call the lib to build the actions directly, and not use `aio app build`. See [this](https://github.com/adobe/aio-cli-plugin-app/blob/74ebd4d396dc8fb91bb193c3e74e7a4fec7e5513/src/lib/build-actions.js#L23).~ We can't do this, since aio-lib-runtime uses [this magic string](https://github.com/adobe/aio-lib-runtime/blob/da90917ef69a02b0131d3abafa123f1b503948ae/src/build-actions.js#L173) and does not expose it
- [x] use aio-lib-runtime to build directly
- [x] debugging via VS Code must work properly (support source maps)
- [x] for Typescript, you will need to npm install (devDep) `typescript, ts-loader` and have a custom `webpack-config.js` (see example project and the [docs](https://developer.adobe.com/app-builder/docs/guides/configuration/webpack-configuration/)). Any includes/excludes should be in `tsconfig.json`. `tsconfig.json` should have the option `sourceMap: true`
- [X] update AB Docs (see PR above)

## How Has This Been Tested?

1. `aio plugins install https://github.com/adobe/aio-cli-plugin-app-dev#story/ACNA-3098`
2. `git clone https://github.com/shazron/test-debugger.git` (from [this repo](https://github.com/shazron/test-debugger))
3. `cd test-debugger`
4. `aio app dev` in the cloned repo folder OR
5. step by step debug the 3 actions in Visual Studio Code via the Debug tab (put breakpoints in the CommonJS, ESM, and TypeScript actions)
6. while the command is still running, make edits to the action code, note the code gets built and updated in the log. Editing and saving while step-by-step debugging is undefined behaviour and not supported.
7. make sure that the edits are reflected in the action results after calling the actions
8. test web-src debugging
9. to uninstall this plugin: `aio plugins uninstall @adobe/aio-cli-plugin-app-dev`

## Testing in your App

1. Update your app's `.vscode/launch.json`, see: https://developer.adobe.com/app-builder/docs/guides/development/#debugging-with-vs-code
2. `aio plugins install https://github.com/adobe/aio-cli-plugin-app-dev#story/ACNA-3098`
3. Add/edit your [webpack-config.js and tsconfig.json](https://github.com/adobe/aio-cli-plugin-app-dev/blob/13cd91c80c35d6376118ec720a1711d67c5035a8/README.md) files
4. `aio app dev` in the your app root OR VS Code debug via the Debug Tab
5. to uninstall this plugin: `aio plugins uninstall @adobe/aio-cli-plugin-app-dev`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/lib/run-dev.js [hunk 1]
```diff
@@ -348,7 +351,10 @@ async function invokeAction ({ actionRequestContext, logger }) {
   let actionFunction
   // catch errors when loading the action function
   try {
-    actionFunction = require(action.function)?.main
+    const actionFolder = path.join(distFolder, packageName, actionName)
+    // TODO: find a better way to get the actual folder name without this magic string
+    const actionPath = `${actionFolder}-temp`
```
Comment (shazron) [inline src/lib/run-dev.js:356]: to not use the magic string, we may have to call the lib to build the actions directly, and not use `aio app build`

Comment (shazron) [reply src/lib/run-dev.js]: We can't do this, since aio-lib-runtime uses [this magic string](https://github.com/adobe/aio-lib-runtime/blob/da90917ef69a02b0131d3abafa123f1b503948ae/src/build-actions.js#L173) and does not expose it

**Review by moritzraho — APPROVED** [review]
lgtm, but didn't test

File: src/lib/actions-watcher.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  return async (filePath) => {
+    if (buildInProgress) {
+      watchLogger.debug(`${filePath} has changed. A build is in progress. This change will be built after completion of current build.`)
+      queue.enqueue(filePath)
+      return
+    }
+
+    buildInProgress = true
+
+    await processChange(filePath)
```
Comment (moritzraho) [inline src/lib/actions-watcher.js:122]: nit ```
let fileToProcess = filePath
do {
      await processChange(fileToProcess)
      fileToProcess = queue.dequeue()
} while (fileToProcess)
```

File: src/lib/actions-watcher.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    // process the rest of the queue, if any
+    let fileToProcess = queue.dequeue()
+    while (fileToProcess) {
+      await processChange(fileToProcess)
+      fileToProcess = queue.dequeue()
+    }
+
```
Comment (moritzraho) [inline src/lib/actions-watcher.js:130]: nit
```suggestion
    let fileToProcess = filePath
    do{
        await processChange(fileToProcess)
        // process the rest of the queue if any
        fileToProcess = queue.dequeue()
    }
    while (fileToProcess)
```

**Comment by purplecabbage** [comment]
Do the actions all have to be built on change, previously they were only ever loaded on demand?
Has this been tested with multiple actions referencing the same code?
ie.
```
actions:
  myaction18open:
    function: actions/myaction/index.js
    web: 'yes'
    runtime: nodejs:18
    inputs:
      LOG_LEVEL: silly
    annotations:
      require-adobe-auth: true
      final: true
  myaction20secure:
    function: actions/myaction/index.js
    web: 'yes'
    runtime: nodejs:20
    inputs:
      LOG_LEVEL: verbose
    annotations:
      require-adobe-auth: false
      final: true
 ```

**Comment by shazron** [comment]
> Do the actions all have to be built on change, previously they were only ever loaded on demand?

If you want the user to potentially wait a few seconds (ok maybe 1 second typically) before they get a response when calling a URL, for webpack to do its thing, we can make it "on-demand". We will not do it everytime of course, but we will need to set a "dirty" flag for that action to signify that a change was done, so we need to webpack first before serving to the user.  This new behavior is exactly what we do for `aio app run` (minus the `deploy` of course - since building in this case is "deploying")

>Has this been tested with multiple actions referencing the same code? 

I just did, it works fine. This is because it will use the module.require cache, like it did previously.

File: src/lib/actions-watcher.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    onChangeHandler,
+    watcher,
+    watcherCleanup
+  }
+}
+
+/**
+ * Simple queue
+ */
+class Queue {
```
[the remark this answers is not in the document]
Comment (shazron) [reply src/lib/actions-watcher.js:62]: I would argue that after seeing the previous action-watcher code in the app-plugin which was hard to parse how it works, there is a major benefit in making the syntax clear in this regard for maintainability purposes (versus .shift() etc which is quite unclear at a glance without going to the docs, since it mutates an array etc). 

Especially since this is not critical path code where a  performance improvement (negligible perhaps, I didn't do a profile) is needed.

**Comment by purplecabbage** [comment]
This pr can be tested directly by running:
`aio plugins install https://github.com/adobe/aio-cli-plugin-app-dev.git\#story/ACNA-3098`

**Comment by shazron** [comment]
@purplecabbage do we need to do anything more for this?

**Comment by purplecabbage** [comment]
In my simple project with 30 actions, this introduces a full minute wait before we accept user input.  Essentially all users are forced into a build step even if their code can simply be run.

There are all kinds of ways that this could be better, for example
- the chance a user wants to call all their actions at once on startup is probably zero, yet we make them wait.  They could be built on-demand, or after the web-server starts up, we could start building in the background, if an action is called before we have built it we would put it next in the build queue.
- use a pre-run hook so devs who need typescript support can have a build step, and the rest of us can still work fast.

**Comment by purplecabbage** [comment]
This breaks hot-reloading of actions, the delete require.cache[temp] is not functioning.
